docs/content/sips/010-key-value.md

```diff
@@ -0,0 +1,118 @@
+// `error::no-such-store` will be raised if the `name` is not recognized.
+open: func(name: string) -> expected<store, error>
+
+// Get the value associated with the specified `key` from the specified
+// `store`.
+//
+// `error::invalid-store` will be raised if `store` is not a valid handle
+// to an open store, and `error::no-such-key` will be raised if there is no
+// tuple for `key` in `store`.
+get: func(store: store, key: string) -> expected<list<u8>, error>
+
+// Set the `value` associated with the specified `key` in the specified
+// `store`, overwriting any existing value.
+//
+// `error::invalid-store` will be raised if `store` is not a valid handle
+// to an open store.
+set: func(store: store, key: string, value: list<u8>) -> expected<unit, error>
+
+// Delete the tuple with the specified `key` from the specified `store`.
+//
+// `error::invalid-store` will be raised if `store` is not a valid handle
+// to an open store.  No error is raised if a tuple did not previously
+// exist for `key`.
+delete: func(store: store, key: string) -> expected<unit, error>
+
+// Return whether a tuple exists for the specified `key` in the specified
+// `store`.
+//
+// `error::invalid-store` will be raised if `store` is not a valid handle
+// to an open store.
+exists: func(store: store, key: string) -> expected<bool, error>
+
+// Close the specified `store`.
+//
+// This has no effect if `store` is not a valid handle to an open store.
+close: func(store: store)
+```
+
+*Note: the pseudo-resource design was inspired by the interface of similar functions in [WASI preview 2](https://github.com/bytecodealliance/preview2-prototyping/blob/d56b8977a2b700432d1f7f84656d542f1d8854b0/wit/wasi.wit#L772-L794).*
+
+#### Built-in local key-value database
+
+Spin will have a built-in database based on SQLite for testing, development, and some production use cases.  We’ve selected SQLite for this purpose because it’s easily embeddable, lightweight, and reliable.  It increases the Spin binary size by about 3% and the release build time by about 1.5%.  We’ve also considered lighter-weight options such as an in-memory hash map which is synced to a flat file on updates, but making this reliably atomic and durable would require reinventing the features SQLite already provides.
+
+By default, each app will have its own, persistent, default store which is independent of all other apps.  This could be implemented as a separate SQLite database for each app, a separate table for each app in a shared database, or even row-level separation via an `app` column in a shared table.  For the initial implementation, we'll use a separate database for each app.  For local apps, the database will be stored by default in a hidden `.spin` directory adjacent to the app's `spin.toml`.  For remote apps (e.g. served by `bindle`), an in-memory database will be used by default.  Both of these defaults may be overridden via a `--key-value-database` option for specifying which file to use.
```

Comment:
Perhaps it would help to rename it to `--default-key-value-file` and document that it only affects the default (i.e. empty name) store, overriding whatever the runtime config says, using the SQLite backend.  I.e. even if the runtime config says use such-and-such Redis server for the default store, this option would force it to use a local SQLite file instead.

Reply:
For the record, we ended up removing the option and will revisit when we have runtime configuration.